### PR TITLE
ENH: 20.2.0 LTS

### PR DESCRIPTION
**Pull request by AndysWorth**
Upgrading to fMRIPrep version 10.2.0 Long Term Support (LTS)

Add a lot of tests

### CODE REVIEW

File: utils/bids/download_run_level.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+                    )
+
+                    bids_path = bids_dir
+                    if Path(bids_dir).exists():
+                        log.info(
+                            "Not actually downloading it because " f"{bids_dir} exists"
+                        )
+                    else:
+                        # only download acquisition data
+                        download_bids_dir(
```
Comment (margaretmahan) [inline utils/bids/download_run_level.py:275]: And as a reminder, download_bids_dir is not in this codebase. You likely have it installed locally which is why you aren't getting errors, but you need to include the bids-client repo in order to access that within a Docker container.

**Review by margaretmahan — CHANGES_REQUESTED** [review]
The flow of the Gear is hard to follow and I've made some comments in some places regarding this. Overall, I recommend explicitly extracting the parameters needed from the config and passing those to downstream functions. And to incorporate list and dictionary comprehensions (or other data structures) to help control the logical flow instead of if/else and for loops, which can become unwieldy. Think about making functions more explicit and modular, and to an extent, independent of the rest of the codebase - and then weave those functions together from run.py as the main flow control point.

The most important aspects I think need to be fixed are as follows:
1. Logging: there should be a lot more `log.info` statements, particularly at logical control points (e.g., calling a function, if/else, etc.) and `log.error` statements where errors occur followed by exiting the Gear immediately, if it's not an error, then `log.warning` or `log.debug`.
2. Set up CircleCI for the tests.
3. Successful run on an instance to take a look at. At least one with fmaps and one without.

Lastly, I think the README isn't formatting correctly. Might want to take a look at that.

File: tests/bin/docker-test.sh [hunk 2]
```diff
@@ -0,0 +1,84 @@
+#!/usr/bin/env sh
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply tests/bin/docker-test.sh:1]: Nice!

File: utils/bids/run_level.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/run_level.py:1]: You're mad python skills are always welcome!

File: utils/bids/run_level.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+"""Determine level at which the gear is running."""
+
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/run_level.py:3]: This is beautiful!  I dropped it in.

File: tests/conftest.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+            for subj in subjects:
+                subj_name = subj.name
+                if (SUBJECTS_DIR / subj_name).exists():
+                    shutil.rmtree(SUBJECTS_DIR / subj_name)
+                print(f"moving subject to {str(SUBJECTS_DIR / subj_name)}")
+                shutil.move(str(subj), str(SUBJECTS_DIR))
+
+    return _method
+
+
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply tests/conftest.py:62]: I think you're right in general about this and I'll work on dropping this habit.  Often I'll test that the value of a printed variable is what is expected which does test the functions a bit.

File: utils/results/zip_htmls.py [hunk 6]
```diff
@@ -1,94 +1,75 @@
...
-import os
 import logging
+import os
 import subprocess as sp
 
-
 log = logging.getLogger(__name__)
 
 
 def zip_it_zip_it_good(context, name):
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/results/zip_htmls.py:12]: I couldn't resist giving it that name.

File: utils/results/zip_intermediate.py [hunk 7]
```diff
@@ -1,96 +1,96 @@
...
+            if ff not in folders_found:
+                log.error(f"Could not find folder '{ff}'")
     else:
-        log.debug('No files or folders specified in config to zip')
+        log.debug("No files or folders specified in config to zip")
 
 
-def zip_all_intermediate_output(context):
+def zip_all_intermediate_output(context, run_label):
     """
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/results/zip_intermediate.py:80]: Yes!  I changed it.

File: utils/results/zip_intermediate.py [hunk 8]
```diff
@@ -1,96 +1,96 @@
...
+import os
 import subprocess as sp
 
-
 log = logging.getLogger(__name__)
 
 
-def zip_intermediate_selected(context):
+def zip_intermediate_selected(context, run_label):
     """
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/results/zip_intermediate.py:11]: I totally re-wrote this as you suggested.

File: utils/results/zip_intermediate.py [hunk 9]
```diff
@@ -1,96 +1,96 @@
-# If you edit this file, please consider updating bids-app-template
+"""Save files from work/, compressed in output/."""
 
-import os
 import logging
+import os
 import subprocess as sp
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/results/zip_intermediate.py:5]: Gone.

File: manifest.json [hunk 10]
```diff
@@ -1,13 +1,13 @@
...
-  "version": "1.0.12_1.5.10",
+  "description": "fMRIPrep 20.2.0 (September 28, 2020) is a functional magnetic resonance imaging (fMRI) data preprocessing pipeline that is designed to provide an easily accessible, state-of-the-art interface that is robust to variations in scan acquisition protocols and that requires minimal user input, while providing easily interpretable and comprehensive error and output reporting. It performs basic processing steps (coregistration, normalization, unwarping, noise component extraction, segmentation, skullstripping etc.) providing outputs that can be easily submitted to a variety of group level analyses, including task-based or resting-state fMRI, graph theory measures, surface or volume-based statistics, etc.",
+  "version": "1.1.7_20.2.0",
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply manifest.json:5]: I already uploaded that version.

File: utils/bids/errors.py [hunk 11]
```diff
@@ -0,0 +1,46 @@
+### BIDS Exceptions
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/errors.py:1]: Done.  This is a hold-over from before the new version was released.

File: Dockerfile [hunk 12]
```diff
@@ -1,37 +1,34 @@
...
 # The last line above is to help keep the docker image smaller
 
-RUN npm install -g bids-validator@1.4.0
+RUN npm install -g bids-validator@1.5.6
 
-RUN pip install flywheel-sdk==11.3.3 \
-        flywheel-bids==0.8.2 \
-        psutil==5.6.3 && \
+COPY requirements.txt /tmp
+RUN pip install -r /tmp/requirements.txt && \
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply Dockerfile:17]: Good idea

File: utils/bids/validate.py [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+                command, stdout=f, stderr=sp.PIPE, universal_newlines=True, check=True
+            )
+
+    except sp.CalledProcessError as err:
+        log.error(repr(err))
+        result = sp.CompletedProcess
+        result.returncode = err.returncode
+
+    msg = command[0] + " return code: " + str(result.returncode)
+    log.info(msg)
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/validate.py:93]: I love f-strings!  This must be older code from a pre- 3.6 python.

File: utils/bids/validate.py [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+
+    # read validation result file to get results as dictionary
+    try:
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/validate.py:96]: I think I ran into a case where there was non-json text in the output.

File: utils/bids/validate.py [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+        for ff in err["files"]:
+            if ff["file"]:
+                err_msg += "      In file " + ff["file"]["relativePath"]
+            if "evidence" in ff and ff["evidence"]:
+                err_msg += ", " + ff["evidence"] + "\n"
+            else:
+                err_msg += "\n"
+        log.error(err_msg)
+
+    # show all warnings
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/validate.py:135]: Yes, that should be the procedure (validate before running the gear).  And fMRIPrep runs bids validation.  So I set the `gear-run-bids-validation` config to False.  This whole thing is probably not necessary at all.  But it is a standard part of the bids-app-template so I hesitate to take it out.  I guess it would save some space.  Hmmmm.  Belt and two pairs of suspenders.

File: utils/bids/validate.py [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+        if err_code > 0:
+            log.exception('Error in BIDS download and validation.')
+            # do not bother processing BIDS data
+
+        else:
+            # process BIDS data...
+
+    See validate_bids() below for an example of calling call_validate_bids().
+"""
+
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/validate.py:38]: I thought that maybe there might be a case when the validator would report something and we could then fix it and still run the gear.  That seems kind of unlikely.

File: utils/bids/download_run_level.py [hunk 17]
```diff
@@ -0,0 +1,364 @@
...
+            acquisition.
+        tree (boolean): create HTML page in output showing 'tree' of bids data
+        tree_title (str): Title to put in HTML file that shows the tree
+        src_data (boolean): download source data (dicoms) as well
+        folders (list): only include the listed folders, if empty include all
+        dry_run (boolean): don't actually download data if True
+        do_validate_bids (boolean): run bids-validator after downloading bids data
+
+    Returns:
+        err_code (int): tells a bit about the error:
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:109]: Good point.  I did it because the validator did it.

File: utils/bids/download_run_level.py [hunk 18]
```diff
@@ -0,0 +1,364 @@
...
+        extra_tree_text += f'  {"source data?":<18}: not downloaded\n'
+    if dry_run:
+        extra_tree_text += f'  {"dry run?":<18}: Yes\n'
+    else:
+        extra_tree_text += f'  {"dry run?":<18}: No\n'
+    extra_tree_text += "\n"
+
+    err_code = 0  # assume no error
+
+    if run_level == "no_destination":
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:149]: For a low-level utility like this, I wanted to be robust to whatever craziness happens and let the caller decide what to do when bad things happen.  I'm trying to think of bids-apps in general for all of the code in the `utils` directory.

File: utils/bids/download_run_level.py [hunk 19]
```diff
@@ -0,0 +1,364 @@
...
+        msg = "Destination does not exist."
+        log.critical(msg)
+        extra_tree_text += f"ERROR: {msg}\n"
+        bids_path = None
+        err_code = 24  # destination does not exist
+
+    else:
+
+        if gtk_context.destination["type"] == "analysis":
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:158]: I assume that bids-apps will want the destination to be an analysis.  This checks that and reports if something else is happening.

File: utils/bids/download_run_level.py [hunk 20]
```diff
@@ -0,0 +1,364 @@
...
+
+        else:
+            log.info(
+                'The destination "%s" is not an analysis or acquisition.',
+                gtk_context.destination["type"],
+            )
+
+        try:  # download BIDS data for the proper run level
+
+            if src_data:
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:180]: Everything in the `utils` directory needs to be general for use in any possible bids-app.

File: run.py [hunk 21]
```diff
@@ -1,345 +1,401 @@
...
+# What level to run at (positional_argument #3)
+ANALYSIS_LEVEL = "participant"
 
-    log = custom_log(context)
+# when downloading BIDS Limit download to specific folders
+DOWNLOAD_MODALITIES = ["anat", "func", "fmap"]  # empty list is no limit
 
-    context.log_config() # not configuring the log but logging the config
+# Whether or not to include src data (e.g. dicoms) when downloading BIDS
+DOWNLOAD_SOURCE = False
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py:42]: These lines at the top of the run script are there so a generic BIDS App can be wrapped in a gear with only slight modification to the gear code.  These are the lines that most likely need to be modified.

File: utils/bids/download_run_level.py [hunk 22]
```diff
@@ -0,0 +1,364 @@
...
+
+            bids_dir = Path(gtk_context.work_dir) / "bids"
+
+            if run_level == "project":
+
+                log.info(
+                    'Downloading BIDS for project "%s"', hierarchy["project_label"]
+                )
+
+                if Path(bids_dir).exists():
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:203]: It happens during testing.  The tests already have the data so it doesn't need to be downloaded.

File: utils/bids/download_run_level.py [hunk 23]
```diff
@@ -0,0 +1,364 @@
...
+            hierarchy["acquisition_label"] = acquisition.label
+            extra_tree_text += (
+                f'  {"acquisition_label":<18}: changed to ' + f"{acquisition.label}\n\n"
+            )
+            run_level = "acquisition"
+
+        else:
+            log.info(
+                'The destination "%s" is not an analysis or acquisition.',
+                gtk_context.destination["type"],
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:175]: `get_analysis_run_level_and_hierarchy()` does not necessarily have to be called before downloading the data here.  This may be called in unforeseen ways.

File: utils/bids/download_run_level.py [hunk 24]
```diff
@@ -0,0 +1,364 @@
...
+                log.info("Downloading source data.")
+            else:
+                log.info("Not downloading source data.")
+
+            if dry_run:
+                log.info("Dry run is set.  No data will be downloaded.")
+            else:
+                log.info("Dry run is NOT set.  Data WILL be downloaded.")
+
+            if len(folders) > 0:
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:190]: This is general to be able to handle any bids app.

File: run.py [hunk 25]
```diff
@@ -1,345 +1,401 @@
...
+BIDS_APP = "fmriprep"
 
-    # Add manifest.json as the manifest_json attribute
-    setattr(context, 'manifest_json', load_manifest_json())
+# What level to run at (positional_argument #3)
+ANALYSIS_LEVEL = "participant"
 
-    log = custom_log(context)
+# when downloading BIDS Limit download to specific folders
+DOWNLOAD_MODALITIES = ["anat", "func", "fmap"]  # empty list is no limit
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py:39]: Keeping this for generality and ease of creating bids-apps.

File: utils/bids/download_run_level.py [hunk 26]
```diff
@@ -0,0 +1,364 @@
...
+                log.info(
+                    'Downloading BIDS for project "%s"', hierarchy["project_label"]
+                )
+
+                if Path(bids_dir).exists():
+                    bids_path = bids_dir
+                    log.info(f"Not actually downloading it because {bids_dir} exists")
+                else:
+                    # don't filter by subject or session, grab all
+                    bids_path = gtk_context.download_project_bids(
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:208]: This is elegant, but it obscures the structure of the code where the data is downloaded at a specific level and what is needed at that level is explicit.

File: utils/bids/download_run_level.py [hunk 27]
```diff
@@ -0,0 +1,364 @@
...
+                            gtk_context.client,
+                            gtk_context.destination["id"],
+                            "acquisition",
+                            bids_dir,
+                            src_data=src_data,
+                            folders=folders,
+                            dry_run=dry_run,
+                        )
+
+            else:
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:285]: I wanted to be very clear about what is going on by including redundant explicit messages because bids app gears can be run in different ways.

File: utils/bids/download_run_level.py [hunk 28]
```diff
@@ -0,0 +1,364 @@
...
+                    # only download data for this session AND this subject
+                    bids_path = gtk_context.download_project_bids(
+                        src_data=src_data,
+                        folders=folders,
+                        dry_run=dry_run,
+                        subjects=[hierarchy["subject_label"]],
+                        sessions=[hierarchy["session_label"]],
+                    )
+
+            elif run_level == "acquisition":
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:249]: It won't.  This is for generality of bids apps.

File: utils/bids/download_run_level.py [hunk 29]
```diff
@@ -0,0 +1,364 @@
...
+            extra_tree_text += f"{bids_err}\n"
+            bids_path = None
+            err_code = 21
+
+        except ApiException as err:
+            log.exception(err, exc_info=True)
+            extra_tree_text += f"EXCEPTION: {err}\n"
+            bids_path = None
+            err_code = 25  # download_bids_dir() ApiException
+
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:306]: The error codes might be useful to the calling code.  I don't want to have a hard exit here -- the calling code should decide what to do.

File: utils/bids/download_run_level.py [hunk 30]
```diff
@@ -0,0 +1,364 @@
...
+            if bidsignore_path:
+                shutil.copy(bidsignore_path, "work/bids/.bidsignore")
+                log.info("Installed .bidsignore in work/bids/")
+
+            try:
+                if do_validate_bids:
+                    # validate (assume returns 1.. something <10 on error)
+                    err_code = validate_bids(bids_path)
+                else:
+                    log.info("Not running BIDS validation")
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:326]: Because sometimes the algorithm can run even when the BIDS is not valid.

File: utils/bids/download_run_level.py [hunk 31]
```diff
@@ -0,0 +1,364 @@
...
+            bids_path = None
+            err_code = 21
+
+        except ApiException as err:
+            log.exception(err, exc_info=True)
+            extra_tree_text += f"EXCEPTION: {err}\n"
+            bids_path = None
+            err_code = 25  # download_bids_dir() ApiException
+
+    if bids_path:  # then validate it
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:307]: They are all needed because `bids-path` can be set in different ways or not set and what is happening must be explicit.

File: utils/bids/download_run_level.py [hunk 32]
```diff
@@ -0,0 +1,364 @@
...
+                else:
+                    log.info("Not running BIDS validation")
+                    err_code = 0
+
+            except Exception as exc:
+                log.exception(exc, exc_info=True)
+                extra_tree_text += f"EXCEPTION: {exc}\n"
+                err_code = 22
+
+        else:  # Nothing was downloaded, so what's the point?
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:334]: It is here to be explicit about how bids data was not downloaded.  That can happen for different reasons.

File: utils/bids/download_run_level.py [hunk 33]
```diff
@@ -0,0 +1,364 @@
...
+
+    if tree:
+        tree_bids(
+            bids_path,
+            str(Path(gtk_context.output_dir) / "bids_tree"),
+            tree_title,
+            extra_tree_text,
+        )
+
+    return err_code
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:364]: That's not for this routine to decide.  The call should decide what to do.

File: run.py [hunk 34]
```diff
@@ -1,345 +1,401 @@
...
 
-        elif context.config['gear-dry-run']:
-            ok_to_run = False
-            result = sp.CompletedProcess
-            result.returncode = 0
-            e = 'gear-dry-run is set: Command was NOT run.'
-            log.warning(e)
-            context.gear_dict['warnings'].append(e)
-            utils.dry_run.pretend_it_ran(context)
+    # Don't run if there were errors or if this is a dry run
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py:249]: The gear shouldn't exit until it has a chance to upload anything that was saved.

File: run.py [hunk 35]
```diff
@@ -1,345 +1,401 @@
...
-            log.info('Command failed.')
+            # Create output directory
+            log.info("Creating output directory %s", output_analysis_id_dir)
+            Path(output_analysis_id_dir).mkdir()
 
-    except Exception as e:
-        context.gear_dict['errors'].append(e)
-        log.critical(e)
-        log.exception('Unable to execute command.')
+            # This is what it is all about
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py:269]: For fMRIPrep, that might be better.  But for general bids apps, I want to exit after uploading whatever might have been saved (like the bids tree output) in case it is helpful to figure out what the gear failed.

File: run.py [hunk 36]
```diff
@@ -1,345 +1,401 @@
...
-        validate_bids(context)
+def main(gtk_context):
 
-    except Exception as e:
-        context.gear_dict['errors'].append(e)
-        log.critical(e)
-        log.exception('Error in BIDS download and validation.',)
+    # Keep a list of errors and warning to print all in one place at end of log
+    # Any errors will prevent the command from running and will cause exit(1)
+    errors = []
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py:179]: The errors are logged as they happen and only printed at the end to get them all in one place.  Waiting to exit after output clean-up at the end will help debug the gear run.

File: run.py [hunk 37]
```diff
@@ -1,345 +1,401 @@
...
-            msg = 'Previous warnings:\n'
-            for err in context.gear_dict['warnings']:
-                if str(type(err)).split("'")[1] == 'str':
-                    # show string
-                    msg += '  Warning: ' + str(err) + '\n'
-                else:  # show type (of warning) and warning message
-                    msg += '  ' + str(type(err)).split("'")[1] + ': ' + str(err) + '\n'
+            log.info("Output directory does not exist so it cannot be removed")
+
+        # save .metadata file
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py:329]: I think it will be used.

File: run.py [hunk 38]
```diff
@@ -1,345 +1,401 @@
...
-from utils.results.zip_intermediate import zip_all_intermediate_output
-from utils.results.zip_intermediate import zip_intermediate_selected
-from utils.results.zip_intermediate import zip_intermediate_selected
-import utils.dry_run
+from utils.results.zip_intermediate import (
+    zip_all_intermediate_output,
+    zip_intermediate_selected,
+)
 
+GEAR = "bids-fmriprep"
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py:29]: To put the main changes necessary to create a bids-app all in one obvious place.

File: run.py [hunk 39]
```diff
@@ -1,345 +1,401 @@
...
+    # 3 positional args: bids path, output dir, 'participant'
+    # This should be done here in case there are nargs='*' arguments
+    # These follow the BIDS Apps definition (https://github.com/BIDS-Apps)
 
-            # don't filter by subject or session, grab all
-            download_bids(context, folders=folders_to_load)
+    # get parameters to pass to the command by skipping gear config parameters
+    # (which start with "gear-").
+    command_parameters = {}
+    for key, val in config.items():
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py:146]: I'm not sure I understand what you mean.  Can you show me an example?

File: utils/bids/download_run_level.py (hunk 1 above)
Comment (AndysWorth) [reply utils/bids/download_run_level.py:275]: All of the tests are run inside the docker container so it must have been installed as a dependency of something else.  But I have added it to `requirements.txt` to be explicit about the version.

File: run.py [hunk 40]
```diff
@@ -1,345 +1,401 @@
...
+                            "value2": "bar",
+                            f"{run_label} {destination_id}": "put this here",
+                        },
+                        "tags": ["ein", "zwei"],
+                    }
+                ],
+                "tags": [run_label, destination_id],
+            },
+        }
+        # with open(f"{gtk_context.output_dir}/.metadata.json", "w") as fff:
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py:371]: This is just temporarily disabled.

File: run.py (hunk 37 above)
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py:329]: I deleted it anyway.

File: tests/unit_tests/test_download_run_level.py [hunk 41]
```diff
@@ -0,0 +1,555 @@
+"""Unit tests for download_run_level.py"""
+
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply tests/unit_tests/test_download_run_level.py:2]: Yes, I need to work on testing the behavior more.  I did some but I also put this on the TODO list.

File: tests/conftest.py [hunk 42]
```diff
@@ -0,0 +1,161 @@
...
+import shutil
+from pathlib import Path
+
+import pytest
+from flywheel_gear_toolkit.utils.zip_tools import unzip_archive
+
+FS_DIR = Path("/usr/local/freesurfer/")
+SUBJECTS_DIR = Path(FS_DIR / "subjects")
+
+
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply tests/conftest.py:12]: I put this on the TODO so the gear can be released sooner.

File: tests/unit_tests/test_download_run_level.py [hunk 43]
```diff
@@ -0,0 +1,555 @@
...
+                tree=False,
+                tree_title=None,
+                src_data=True,
+                folders=[],
+                dry_run=True,
+            )
+
+    assert len(caplog.records) == 6
+    assert "No BIDS data was found" in caplog.records[4].message
+
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply tests/unit_tests/test_download_run_level.py:548]: Yup

File: utils/fly/make_file_name_safe.py [hunk 44]
```diff
@@ -1,40 +1,35 @@
...
 
-
 log = logging.getLogger(__name__)
 
 
-def make_file_name_safe(input_basename, replace_str=''):
-    """
-    removes non-safe characters from a filename and returns a filename with these characters replaced with replace_str
+def make_file_name_safe(input_basename, replace_str=""):
+    """Remove non-safe characters from a filename and return a filename with
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/fly/make_file_name_safe.py:10]: This is old code and I think David wrote it.  It used to be in the gear-toolkit and should be there again.  I created an issue to add it [#6](https://gitlab.com/flywheel-io/public/gear-toolkit/-/issues/6).

File: run.py [hunk 45]
```diff
@@ -1,345 +1,401 @@
...
+    exec_command,
+)
+from flywheel_gear_toolkit.licenses.freesurfer import install_freesurfer_license
+from flywheel_gear_toolkit.utils.zip_tools import zip_output
+
+from utils.bids.download_run_level import download_bids_for_runlevel
+from utils.bids.run_level import get_run_level_and_hierarchy
+from utils.dry_run import pretend_it_ran
+from utils.fly.make_file_name_safe import make_file_name_safe
 from utils.results.zip_htmls import zip_htmls
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py]: Gone

File: manifest.json [hunk 46]
```diff
@@ -84,7 +84,7 @@
       "default": false,
       "type": "boolean"
     },
-    "boilerplate": {
+    "boilerplate_only": {
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply manifest.json:87]: I followed the order they used [here](https://fmriprep.org/en/stable/usage.html#command-line-arguments).  I assume they did it in that order for a reason.  Ordering will be maintained in gears v3.

File: utils/bids/download_run_level.py [hunk 47]
```diff
@@ -0,0 +1,364 @@
+#!/usr/bin/env python3
+"""A robust template for accessing BIDS formatted data."""
+
+import json
+import logging
+import shutil
+from pathlib import Path
+
+from flywheel import ApiException
+from flywheel_bids.export_bids import download_bids_dir
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py:10]: I added it to requirements.txt

File: utils/bids/download_run_level.py [hunk 48]
```diff
@@ -0,0 +1,364 @@
...
+
+from utils.bids.errors import BIDSExportError
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply utils/bids/download_run_level.py]: Done.

File: run.py [hunk 49]
```diff
@@ -1,345 +1,401 @@
...
+            gtk_context,
+            hierarchy,
+            tree=tree,
+            tree_title=tree_title,
+            src_data=DOWNLOAD_SOURCE,
+            folders=DOWNLOAD_MODALITIES,
+            dry_run=dry_run,
+            do_validate_bids=config.get("gear-run-bids-validation"),
+        )
+        if error_code > 0 and not config.get("gear-ignore-bids-errors"):
```
[the remark this answers is not in the document]
Comment (AndysWorth) [reply run.py:241]: This is true for fMRIPrep but I want to keep the structure of run.py the same for bids apps in general.

**Comment by AndysWorth** [comment]
Great and thorough review!  Thanks for all of the improvements that are now in the code.  I didn't do some of your suggestions because I want to keep all of the BIDS App gears in sync with the `bids-app-template`.  There are a lot of things that are done to be generally useful in BIDS Apps for both existing and future gears.

I left one comment unresolved -- I'd like to get a better idea about what you are suggesting and why.

**Comment by AndysWorth** [comment]
More changes will be made later.
